master/internal/core_checkpoint_upstream_intg_test.go

```diff
@@ -0,0 +1,241 @@
+	zr, err := zip.NewReader(strings.NewReader(content), int64(len(content)))
+	require.NoError(t, err, "failed to create a zip reader")
+	gotMap := make(map[string]string)
+	prefix := S3_TEST_PREFIX + "/" + id + "/"
+	for _, f := range zr.File {
+		pathname := strings.TrimPrefix(f.Name, prefix)
+		buf := &strings.Builder{}
+		rc, err := f.Open()
+		require.NoError(t, err, "unable to decompress file", f.Name)
+		_, err = io.Copy(buf, rc)
+		require.NoError(t, err, "unable to read content of file", f.Name)
+		rc.Close()
+		gotMap[pathname] = buf.String()
+	}
+	require.Equal(t, mockCheckpointContent, gotMap)
+}
+
+func addMockCheckpointDB(t *testing.T, pgDB *db.PgDB, id uuid.UUID) {
+	etc.SetRootPath(db.RootFromDB)
+	user := db.RequireMockUser(t, pgDB)
+	// Using a different path than DefaultTestSrcPath since we are one level up than most db tests
+	exp := mockExperimentS3(t, pgDB, user, "../../examples/tutorials/mnist_pytorch")
+	tr := db.RequireMockTrial(t, pgDB, exp)
+	allocation := db.RequireMockAllocation(t, pgDB, tr.TaskID)
+	// Create checkpoints
+	checkpoint := db.MockModelCheckpoint(id, tr, allocation)
+	err := pgDB.AddCheckpointMetadata(context.TODO(), &checkpoint)
+	require.NoError(t, err)
+}
+
+func createCheckpoint(t *testing.T, pgDB *db.PgDB) (string, error) {
+	id := uuid.New()
+	addMockCheckpointDB(t, pgDB, id)
+	err := createMockCheckpointS3(S3_TEST_BUCKET, S3_TEST_PREFIX+"/"+id.String())
+	return id.String(), err
+}
+
+func newEchoContext() (echo.Context, *httptest.ResponseRecorder) {
+	e := echo.New()
+	rec := httptest.NewRecorder()
+	return e.NewContext(nil, rec), rec
+}
+
+func TestGetCheckpointEcho(t *testing.T) {
+	var id string
+	cases := []struct {
+		DenyFuncName string
+		IDToReqCall  func(id string) error
+		Params       []any
+	}{
+		{"CanGetCheckpointTgz", func(id string) error {
+			api, _, _ := SetupCheckpointTestEcho(t)
+			id, err := createCheckpoint(t, api.m.db)
+			if err != nil {
+				return err
+			}
+			ctx, rec := newEchoContext()
+			ctx.SetParamNames("checkpoint_uuid")
+			ctx.SetParamValues(id)
+			ctx.SetRequest(httptest.NewRequest(http.MethodGet, "/tgz", nil))
+			err = api.m.getCheckpointTgz(ctx)
+			checkTgz(t, rec.Body, id)
+			return err
+		}, []any{mock.Anything, mock.Anything}},
+		{"CanGetCheckpointZip", func(id string) error {
+			api, _, _ := SetupCheckpointTestEcho(t)
+			id, err := createCheckpoint(t, api.m.db)
+			if err != nil {
+				return err
+			}
+			ctx, rec := newEchoContext()
+			ctx.SetParamNames("checkpoint_uuid")
+			ctx.SetParamValues(id)
+			ctx.SetRequest(httptest.NewRequest(http.MethodGet, "/zip", nil))
+			err = api.m.getCheckpointZip(ctx)
+			checkZip(t, rec.Body.String(), id)
+			return err
+		}, []any{mock.Anything, mock.Anything}},
+	}
+
+	for _, curCase := range cases {
+		require.NoError(t, curCase.IDToReqCall(id))
+	}
+}
+
+func mockExperimentS3(t *testing.T, pgDB *db.PgDB, user model.User, folderPath string) *model.Experiment {
```

Comment:
This was copied from `RequireMockExperiment()`. Let me know if you prefer reusing it.